### PR TITLE
Add More Motion

### PR DESCRIPTION
**Pull request by NexusKitten**
More Motion is an extension to add several utility blocks pertaining to the "Motion" category, such as blocks for distance, stepping, and checking position. Although it's blocks are somewhat limited as of currently, I wanted to set the groundwork for a motion extension since we don't currently have one yet.

image (30)

### CODE REVIEW

**Comment by idoinfactexist** [comment]
call it motion plus, its just too easy

**Comment by horizontal-shading** [comment]
YESSSS MOTION+ I WAS WAITING FOR THIS

**Comment by Martinelplayz** [comment]
(distance between (x1) (y1) and (x2) (y2))

**Comment by NexusKitten** [comment]
> (distance between (x1) (y1) and (x2) (y2))

This block is already in Graphics 2D.

**Comment by CubesterYT** [comment]
@NexusKitten are you ok lol?

**Comment by GarboMuffin** [comment]
please keep the pull request names reasonable if you want them to be looked at

**Comment by NexusKitten** [comment]
> please keep the pull request names reasonable if you want them to be looked at.

Sorry about that!

File: extensions/NexusKitten/moremotion.js [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+      } else {
+        util.target.setDirection(((180 / Math.PI) * Math.atan((x - util.target.x) / (y - util.target.y))));
+      }
+    }
+
+    rotationStyle(args, util) {
+    return util.target.rotationStyle;
+    }
+
+    fence(util) {
```
Comment (GarboMuffin) [inline extensions/NexusKitten/moremotion.js:220]: - first argument is always args (an empty object in this case). second argument is always util.
 - you should make the fencing match Scratch. here's a hint: https://github.com/TurboWarp/scratch-vm/blob/9329427fb57de83cf5d8c2774b55013c60f39ed4/src/sprites/rendered-target.js#L269

File: extensions/NexusKitten/moremotion.js [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+      // Shoutout to Pythagoras!
+      return Math.sqrt(((x - util.target.x) ** 2) + ((y - util.target.y) ** 2));
+    }
+
+    steptowards(args, util) {
+      const x = Scratch.Cast.toNumber(args.X);
+      const y = Scratch.Cast.toNumber(args.Y);
+      const steps = Scratch.Cast.toNumber(args.STEPS);
+      const val = steps / (Math.sqrt(((x - util.target.x) ** 2) + ((y - util.target.y) ** 2)));
+      util.target.setXY(((x - util.target.x) * (val)) + util.target.x, ((y - util.target.y) * (val)) + util.target.y);
```
Comment (GarboMuffin) [inline extensions/NexusKitten/moremotion.js:246]: If the distance to the point is less than the distance to step, wouldn't it make sense to not step too far?

File: extensions/NexusKitten/moremotion.js [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+                type: Scratch.ArgumentType.NUMBER,
+                defaultValue: '0'
+              }
+            }
+          },
+          {
+            filter: [Scratch.TargetType.SPRITE],
+            opcode: 'inbetween',
+            blockType: Scratch.BlockType.BOOLEAN,
+            text: 'in-between x1: [X1] y1: [Y1] x2: [X2] y2: [Y2]?',
```
Comment (GarboMuffin) [inline extensions/NexusKitten/moremotion.js:169]: a name like "in rectangle" or something might be less ambiguous

File: extensions/NexusKitten/moremotion.js [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+              Y2: {
+                type: Scratch.ArgumentType.NUMBER,
+                defaultValue: '100'
+              }
+            }
+          },
+        ],
+        menus: {
+          WHAT: {
+            acceptreporters: false,
```
Comment (GarboMuffin) [inline extensions/NexusKitten/moremotion.js:192]: why not true?

File: extensions/NexusKitten/moremotion.js (hunk 1 above)
Comment (NexusKitten) [reply extensions/NexusKitten/moremotion.js]: All issues you presented should be resolved in the new commit.

**Comment by BlueDome77** [comment]
It should just be called **Motion Plus** to match the other category expansion extensions

**Comment by GarboMuffin** [comment]
I made some changes

 - distinction between costume size and actual sprite size after rotation, etc.
 - touching point works with rotation etc. and checks if the costume actually exists at that point
 - rectangle touching block now works with rotation etc. and checks if any part of the costume actually intersects the rectangle (it also fixes the argument order if someone gets it backwards)

the last 2 are significant breaking changes, so let me know if you need the old behavior back

**Comment by LilyMakesThings** [comment]
"touching" as a concept is a sensing thing. Wouldn't it make more sense within sensing+?

**Comment by NexusKitten** [comment]
> "touching" as a concept is a sensing thing. Wouldn't it make more sense within sensing+?
> 
> "Motion" describes movement, and the reporters in the motion category typically get an axis of movement, not a condition. Otherwise, distance would've been in motion.

this is true, I mostly added it into a motion because:
1. both touching blocks are focused on the sprites position, and I figured from a logical standpoint people would associate position booleans with motion instead of sensing
2. sensing+ is kind of a mess of blocks right now, I'd rather not throw it further into "utilities" category until we can find a solution to its current problems

**Pull request merged**